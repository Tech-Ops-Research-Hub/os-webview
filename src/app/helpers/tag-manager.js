/* eslint-disable camelcase, no-nested-ternary */
import {accountsModel} from '~/models/usermodel';
import {getPageDescription} from '~/helpers/use-document-head';
const tagManagerID = 'GTM-W6N7PB';

window.dataLayer = window.dataLayer || [];

// eslint-disable-next-line max-params
(function (w, d, s, l, i) {
    // Disable ESLint rules since we're copying Google's script
    /* eslint-disable one-var, prefer-const, prefer-template */
    w[l] = w[l] || [];
    w[l].push({
        'gtm.start': new Date().getTime(),
        event: 'gtm.js'
    });

    let f = d.getElementsByTagName(s)[0],
        j = d.createElement(s),
        dl = l !== 'dataLayer' ? '&l=' + l : '';

    // Breaks in tests because there are no scripts
    if (f) {
        j.async = true;
        j.src = '//www.googletagmanager.com/gtm.js?id=' + i + dl;
        f.parentNode.insertBefore(j, f);
    }
})(window, document, 'script', 'dataLayer', tagManagerID);

/*
 * google analytics 4 has enhanced tracking for page views when javascript
 * navigates with the history api, in these cases GA4 seems to wait for the
 * page metadata to be updated so it gets the right title, but the initial
 * page load fires as soon as the config is loaded, so we delay the
 * app_initialized event until the page title has been updated
 */
const initialTitle = document.title;
const initialDescription = getPageDescription();

let timeout;
let interval;
const dataInitialized = Promise.race([
  new Promise((resolve) => {
    timeout = window.setTimeout(resolve, 1000);
  }),
  new Promise((resolve) => {
   interval = window.setInterval(() => {
      if (initialTitle !== document.title && initialDescription !== getPageDescription()) {
        resolve();
      }
    }, 1);
  })
])
  .finally(() => {
    window.clearInterval(interval);
    window.clearTimeout(timeout);
  });

Promise.all([accountsModel.load(), dataInitialized]).then(([accountResponse]) => {
    const role = ['instructor', 'student'].includes(accountResponse.self_reported_role) ?
        accountResponse.self_reported_role :
        accountResponse.uuid ? 'other' : 'none';
    const roleTag = `role=${role}`;

    const faculty = accountResponse.faculty_status;
    const facultyTag = faculty ? `faculty=${faculty}` : undefined;

    const usingOpenstax = accountResponse.using_openstax;
    const usingOpenstaxTag = usingOpenstax ? 'adopter=yes' : undefined;

    const user_tags = ['', roleTag, facultyTag, usingOpenstaxTag, '']
      .filter((x) => x !== undefined).join(',');

    window.dataLayer.push({event: 'app_loaded', app: 'osweb', user_tags});
});

export function setContentTags(tags) {
    const content_tags = tags.length > 0 ?
        ['', ...tags, ''].join(',') :
        undefined;

    if (content_tags) {
        window.dataLayer.push({event: 'app_config', content_tags});
    }
}
