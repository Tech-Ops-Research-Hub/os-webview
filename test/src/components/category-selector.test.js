import CategorySelector from '~/components/category-selector/category-selector';
import {clickElement} from '../../test-utils';
import $ from '~/helpers/$';

describe('CategorySelector', () => {
    const p = new CategorySelector();

    it('creates', () => {
        expect(p.el.innerHTML).toBeTruthy();
    });
    it('handles math click', () => {
        const mathButton = p.el.querySelector('[data-value="math"]');

        clickElement(mathButton);
        expect(mathButton.getAttribute('aria-pressed')).toBe('true');
        expect(p.model.isSelected('math')).toBe(true);
    });
    it('handles selection by key enter', () => {
        const scienceButton = p.el.querySelector('[data-value="science"]');

        scienceButton.dispatchEvent(
            new KeyboardEvent('keydown', {
                key: 'Enter',
                code: $.key.enter,
                keyCode: $.key.enter,
                bubbles: true,
                cancelable: true
            })
        );
        expect(p.model.isSelected('science')).toBe(true);
    });
    it('matches snapshot', () => {
        console.log('To update snapshot: node_modules/.bin/jest --updateSnapshot --testNamePattern=Category');
        expect(p.el.innerHTML).toMatchSnapshot();
    });
});
