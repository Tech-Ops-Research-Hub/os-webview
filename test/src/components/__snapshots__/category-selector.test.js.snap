// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategorySelector matches snapshot 1`] = `"<div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"view-all\\" aria-pressed=\\"false\\"><span data-html=\\"items[0].html\\">View All</span></div><div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"math\\" aria-pressed=\\"false\\"><span data-html=\\"items[1].html\\">Math</span></div><div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"science\\" aria-pressed=\\"true\\"><span class=\\"on-mobile filter-by\\">Filter by: </span><span data-html=\\"items[2].html\\">Science</span><span class=\\"on-mobile fa fa-chevron-down\\"></span></div><div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"social-sciences\\" aria-pressed=\\"false\\"><span data-html=\\"items[3].html\\">Social Sciences</span></div><div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"humanities\\" aria-pressed=\\"false\\"><span data-html=\\"items[4].html\\">Humanities</span></div><div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"business\\" aria-pressed=\\"false\\"><span data-html=\\"items[5].html\\">Business</span></div><div class=\\"filter-button\\" tabindex=\\"0\\" role=\\"button\\" data-value=\\"ap\\" aria-pressed=\\"false\\"><span data-html=\\"items[6].html\\">AP<sup>Â®</sup></span></div>"`;
